build.rs

```diff
@@ -27,4 +27,36 @@ fn main() {
     }
     let target = env::var("TARGET").unwrap();
     println!("cargo:rustc-env=TARGET={target}");
+
+    let target_os = env::var("CARGO_CFG_TARGET_OS");
+    let target_env = env::var("CARGO_CFG_TARGET_ENV");
+    if Ok("windows") == target_os.as_deref() && Ok("msvc") == target_env.as_deref() {
+        // # Only search system32 for DLLs
+        //
+        // This applies to DLLs loaded at load time. However, this setting is ignored
+        // before Windows 10 RS1.
+        println!("cargo:cargo:rustc-link-arg-bin=rustup-init=/DEPENDENTLOADFLAG:0x800");
+
+        // # Delay load
+        //
+        // Delay load dlls that are not "known DLLs".
+        // Known DLLs are always loaded from the system directory whereas other DLLs
+        // are loaded from the application directory. By delay loading the latter
+        // we can ensure they are instead loaded from the system directory.
+        //
+        // This will work on all supported Windows versions but it relies on
+        // using `SetDefaultDllDirectories` before any libraries are loaded.
+        let delay_load_dlls = ["bcrypt", "powrprof", "secur32"];
+        for dll in delay_load_dlls {
+            println!("cargo:rustc-link-arg-bin=rustup-init=/delayload:{dll}.dll");
+        }
+        println!("cargo:rustc-link-arg-bin=rustup-init=delayimp.lib");
+
+        // # Turn linker warnings into errors
+        //
+        // Rust hides linker warnings meaning mistakes may go unnoticed.
```

Comment:
Are there generally good reasons to hide these warnings "upstream" (I guess in rustc)? Or is there a good reason that rustup's behavior, in particular, should be different from the default?

Reply:
Personally I think rustc should not be hiding linker warnings but I'm not trying to litigate that here. However, the reason I think it's especially important in this case is that we're directly and explicitly setting some linker arguments and it would not be good if problems with them (e.g. typos) were being silently ignored.